### PR TITLE
Optimise recalcBlockCounts

### PR DESCRIPTION
**Pull request by AlfieC**
Minimises operations and cuts out bit shifting & 2 loops.

### CODE REVIEW

**Comment by Techcable** [comment]
Please actually make the changes directly instead of commenting out the old methods.

**Comment by AlfieC** [comment]
They're left there to make updating easier - if Mojang change this method completely, it's useful to see the original. It can be removed on merge.

**Comment by Techcable** [comment]
If mojangs makes a change the diff won't trigger a conflict, causing us to miss changes on update.
Additionally it's more difficult for maintainers to know what changes the patch actually made, which wastes time and effort in a update.
Git is smart enough to automatically resolve simple conflicts on an update, but this makes it ignore changes by mojang, causing bugs on update.
We'll already be able to see the old mojang code, since git shows it to us if there's a conflict.

**Comment by AlfieC** [comment]
I only left it commented instead of removing as that's the style I saw in the `Disable ticking snow blocks` patch. Maybe that's different as it's a constructor that was removed - but I didn't want to decide for myself.

**Comment by AlfieC** [comment]
Updated PR to remove old method instead of commenting, and added an extra optimisation.

File: Spigot-Server-Patches/0181-Optimise-recalcBlockCounts.patch [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+--- a/src/main/java/net/minecraft/server/DataPaletteBlock.java
++++ b/src/main/java/net/minecraft/server/DataPaletteBlock.java
+@@ -17,6 +17,14 @@ public class DataPaletteBlock implements DataPaletteExpandable {
+     private static int b(int i, int j, int k) {
+         return j << 8 | k << 4 | i;
+     }
++    
++    // Paper start
++    public IBlockData getBlockData(int position) {
++        IBlockData iblockdata = this.c.a(this.b.a(position));
```
Comment (aikar) [inline Spigot-Server-Patches/0181-Optimise-recalcBlockCounts.patch]: This absolutely needs obfuscation helpers

File: Spigot-Server-Patches/0181-Optimise-recalcBlockCounts.patch [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+         IBlockData iblockdata1 = this.getType(i, j, k);
+         Block block = iblockdata1.getBlock();
+         Block block1 = iblockdata.getBlock();
+-
+-        if (block != Blocks.AIR) {
+-            --this.nonEmptyBlockCount;
+-            if (block.isTicking()) {
+-                --this.tickingBlockCount;
++        // Paper start
++        if (block != block1) {
```
Comment (aikar) [inline Spigot-Server-Patches/0181-Optimise-recalcBlockCounts.patch]: Based on what this change is doing, please undo the indentation change of all of this inner code.

I'm almost willing to say just short circuit return and copy the last line that isn't part of the if(), but that would introduce risk for bugs in future.

But im more comfortable with the inner code changing. and being ok still.

**Review by aikar — CHANGES_REQUESTED** [review]

**Comment by aikar** [comment]
@AlfieC can you get the requested changes in and ensure this applies over latest?

**Comment by AlfieC** [comment]
PR updated.

**Comment by kashike** [comment]
Do we want this?

**Pull request closed without merging**